Expose `UnrealClassRegistrar.h` functions to python API

RachithP opened the pull request.

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -181,6 +184,547 @@ class GameWorldService {
+                return reinterpretAs<uint64_t>(UnrealClassRegistrar::createSceneComponentOutsideOwnerConstructor(class_name, reinterpretAs<USceneComponent*>(owner), name));
+            });
+
+        unreal_entry_point_binder->bindFuncUnreal("game_world_service", "create_new_object",
```
No good reason to deviate from the underlying Unreal function name here.

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -27,6 +28,8 @@
 // data types used in SpCore cannot be sent via RPCLib, e.g., FProperty* in PropertyDesc.
 
 MSGPACK_ADD_ENUM(EIncludeSuperFlag::Type);
+MSGPACK_ADD_ENUM(ELoadFlags);
```
Create matching Python enums for all of these enums.

mikeroberts3000 replied:
Update: I think I have a cleaner way of doing synchronizing these enums that doesn't involve these `MSGPACK` macros. See comment below.

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -181,6 +184,547 @@ class GameWorldService {
+                        reinterpretAs<UPackage*>(external_package)));
+            });
+
+        unreal_entry_point_binder->bindFuncUnreal("game_world_service", "load_object",
```
Let's include `get_static_class` here too.

mikeroberts3000 replied:
(Order still not matched to `UnrealClassRegistrar`)

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -181,6 +184,547 @@ class GameWorldService {
+                        reinterpretAs<UObject*>(outer),
+                        Unreal::toFName(name),
+                        flags,
+                        reinterpretAs<UObject*>(uobject_template),
```
`reinterpret_cast`

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -181,6 +184,547 @@ class GameWorldService {
+
+        unreal_entry_point_binder->bindFuncUnreal("game_world_service", "get_children_components_by_tag_as_map_from_scene_component",
+            [this](const std::string& class_name, const uint64_t& parent, const std::string& tag, const bool& include_all_descendants) -> TMapStringToInt {
+                return getIntMapFrom(UnrealClassRegistrar::getChildrenComponentsByTagAsMap(class_name, reinterpretAs<USceneComponent*>(parent), tag, include_all_descendants));
```
`toMap`

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -181,6 +184,547 @@ class GameWorldService {
             [this](const uint64_t& instance) -> uint64_t {
                 return reinterpret_cast<uint64_t>(reinterpret_cast<UObject*>(instance)->GetClass());
             });
+
```
`const uint64_t& uobject`

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -181,6 +184,547 @@ class GameWorldService {
             [this](const uint64_t& instance) -> uint64_t {
```
`const uint64_t& uobject`

mikeroberts3000 commented on cpp/unreal_plugins/SpCore/Source/SpCore/UnrealClassRegistrar.cpp:
```diff
@@ -75,6 +75,22 @@ void UnrealClassRegistrar::terminate()
+}
+
+//
+// Get Default object using a class name instead of template parameters
```
`Get default object using a class name instead of template parameters`

mikeroberts3000 commented on cpp/unreal_plugins/SpCore/Source/SpCore/UnrealClassRegistrar.h:
```diff
@@ -42,6 +42,18 @@ class SPCORE_API UnrealClassRegistrar
+    // Get Default object using a class name instead of template parameters
+    //
+
+    static UObject* getDefaultObject(const std::string& class_name);
```
Let's order these functions as follows:

```
newObject
loadObject
getDefaultObject
getStaticClass
```

I think this is a slightly more logical ordering, because then the return types of of all our create/get functions will be sorted roughly in high-level-to-low-level order (`AActor`, `UComponent`, `UObject`, `UClass`).

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -36,6 +39,107 @@ struct GameWorldServicePropertyDesc
     MSGPACK_DEFINE_MAP(property_, value_ptr_);
 };
 
+/* Struct of optional parameters passed to SpawnActor function(s). */
```
I'm surprised this struct doesn't cause a naming conflict with the existing Unreal struct? If I refer to `FActorSpawnParameters` in a block of code, how does the compiler know which struct I'm referring to? I think we should call this struct `FSpActorSpawnParameters` to avoid confusion. And for brevity, we should just provide the following top-level comment and remove the comments, which are not formatted correctly.

```
// This struct is intended to be identical to Unreal's FActorSpawnParameters struct, except we tag our struct
// as a USTRUCT so it is easier to pass to and from Python as a JSON string, see:
//     Engine/Source/...
```

mikeroberts3000 replied:
Updates:

- An inheritance-based solution doesn't work, even for trivial structs. The Unreal Build Tool generates errors because it wants the base struct to also be a `USTRUCT`.
- A composition-based solution will compile, but the problem is that the raw inner struct cannot be tagged as a `UPROPERTY` because its type is not a `USTRUCT`. Therefore, it will not be exposed to Unreal's property system, and by extension will not be exposed to our JSON serialization/deserialization functions. But if the inner struct cannot be set with JSON, then it is not useful in our setting.
- Based on these results, it looks like we'll need to do things the ugly verbose way. Note that everything in our `USTRUCT` will need to be declared as a `UPROPERTY`.

RachithP replied:
Update:
Modified the code to do this the verbose way - defining a custom struct with USTRUCT() macro, where each property in this new struct corresponds to (or is a duplicate for) a property from the Unreal's struct.

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -36,6 +39,107 @@ struct GameWorldServicePropertyDesc
+    /** Determines whether to multiply or override root component with provided spawn transform */
+    ESpawnActorScaleMethod TransformScaleMethod = ESpawnActorScaleMethod::MultiplyWithRoot;
+
+private:
```
Don't need these friend definitions.

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -36,6 +39,107 @@ struct GameWorldServicePropertyDesc
+
+public:
+
+    bool IsRemoteOwned() const { return bRemoteOwned; }
```
Don't need this function.

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -36,6 +39,107 @@ struct GameWorldServicePropertyDesc
+    APawn* Instigator;
+
+    /* The ULevel to spawn the Actor in, i.e. the Outer of the Actor. If left as NULL the Outer of the Owner is used. If the Owner is NULL the persistent level is used. */
+    class	ULevel* OverrideLevel;
```
```
ULevel* OverrideLevel;
```

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -178,8 +282,617 @@ class GameWorldService {
+
+        // TODO: How do we expose FActorSpawnParameters?
+        unreal_entry_point_binder->bindFuncUnreal("game_world_service", "spawn_actor",
+            [this](const std::string& class_name, const std::vector<double>& location, const std::vector<double>& rotation, const spawn_params) -> uint64_t {
```
How does this compile? What is the type of `spawn_params`?

RachithP replied:
aah, my bad, this functionality should have been commented when I pushed the changes.

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -178,8 +282,617 @@ class GameWorldService {
+                        world_,
+                        FVector(location.at(0), location.at(1), location.at(2)),
+                        FRotator(rotation.at(0), rotation.at(1), rotation.at(2)),
+                        FActorSpawnParameters()));
```
When you're ready, you can convert a string to an FSpActorSpawnParameters as follows:

```
FSpActorSpawnParameters sp_actor_spawn_parameters;
Unreal::setObjectPropertiesFromString(&sp_actor_spawn_parameters, spawn_params)
```

From here, I'm not sure what the cleanest way to convert between `FSpActorSpawnParameters` and `FActorSpawnParameters`, aside from the obvious ugly solution of copying every member. Doing a `memcpy` would work if we lay out all of the data in exactly the same order as the underlying Unreal struct, but this is brittle because the Unreal developers might change the data layout and we wouldn't notice.

Can we use inheritance to avoid the repetition? In other words, can we just make our `FSpActorSpawnParameters` inherit from `FActorSpawnParameters`? That would be ideal, because then we wouldn't need to explicitly duplicate the data layout.

mikeroberts3000 replied:
Update: no we can't use inheritance (or composition).

mikeroberts3000 commented on python/spear/game_world_service.py:
```diff
@@ -2,9 +2,92 @@
+        property_
+        value_ptr_
+
+    # should match the EIncludeSuperFlag::Type enum in Engine\Source\Runtime\CoreUObject\Public\UObject\Class.h
```
forward slashes for consistency

mikeroberts3000 replied:
See my most recent design sketch to avoid duplicating most of these enums in Python.

mikeroberts3000 commented on python/spear/game_world_service.py:
```diff
@@ -2,9 +2,92 @@
 
 class GameWorldService():
+
+    class PropertyDesc():
```
add comment that this enum needs to match its C++ counterpart

mikeroberts3000 replied:
See my most recent design sketch to avoid duplicating most of these enums in Python.

mikeroberts3000 commented on python/spear/game_world_service.py:
```diff
@@ -2,9 +2,92 @@
+        ExcludeSuperFlag = 0
+        IncludeSuperFlag = 1
+
+    # should match the ELoadFlags class in Engine\Source\Runtime\CoreUObject\Public\UObject\ObjectMacros.h
```
no need to repeat all of these comments, since we're already linking to the Unreal source file

mikeroberts3000 replied:
See my most recent design sketch to avoid duplicating most of these enums in Python.

mikeroberts3000 commented on examples/mujoco_interop/run.py:
```diff
@@ -45,12 +45,11 @@ def unreal_rpy_from_mujoco_quaternion(mujoco_quaternion):
 
     unreal_set_actor_location_and_rotation_func = spear_instance.game_world_service.find_function_by_name(
-        uclass=unreal_actor_class, name="K2_SetActorLocationAndRotation", include_super_flag=1)
+        uclass=unreal_actor_static_class, name="K2_SetActorLocationAndRotation", include_super_flag=spear_instance.game_world_service.EIncludeSuperFlag.IncludeSuperFlag.value)
```
Great.

mikeroberts3000 replied:
In my new design sketch, this would be replaced with `include_super_flag="IncludeSuperFlag"`.

mikeroberts3000 commented on examples/mujoco_interop/run.py:
Let's rename `mj_body_xpos` and `mj_body_quat` to `mj_bodies_xpos` and `mj_bodies_quat`. I named these variables poorly in my previous review comments.

mikeroberts3000 commented:
This PR is looking great. None of the changes I suggested are super complicated, but they might touch a lot of the new code, so I'd like to have one more look before approving.

mikeroberts3000 commented on python/spear/game_world_service.py:
```diff
@@ -2,9 +2,92 @@
+	# while checking the object flags is much faster outside of it where the Object pointer is already available and most likely cached.
+	RF_AllocatedInSharedPage	=0x80000000,	///< Allocated from a ref-counted page shared with other UObjects
+
     def __init__(self, rpc_client):
```
- rename `func_args` to `args` in `call_function(...)`

mikeroberts3000 commented:
Here is a design sketch for how I think we should synchronize Unreal enums across Python and C++. Suppose there is some underlying Unreal enum that is not tagged as a `UENUM` that we would like to be able to easily pass to and from Python.

```
enum class UnrealEnum
{
    Zero = 0,
    One = 1,
    Two = 2
};
```

Our first step is to declare a new `UENUM` with matching values. No matter what approach we use, we will need to declare an extra enum somewhere. I think it's better to do this in C++ rather than Python, because we can avoid explicitly referring to the raw underlying integer values.

```
UENUM()
enum class ESpEnum
{
    Zero = static_cast<std::underlying_type_t<UnrealEnum>>(UnrealEnum::Zero),
    One = static_cast<std::underlying_type_t<UnrealEnum>>(UnrealEnum::One),
    Two = static_cast<std::underlying_type_t<UnrealEnum>>(UnrealEnum::Two)
};
```

Our next step is to declare a `USTRUCT` that contains our `UENUM`. We don't need to do anything similar in the `MSGPACK` approach, so this is a legitimate downside of my proposed approach. But it will provide a lot of extra convenience.

```
USTRUCT()
struct FSpEnum
{
    GENERATED_BODY()

    UPROPERTY()
    ESpEnum SpEnum;
};
```

Now suppose we want to call an Unreal function from a service entry point, where the Unreal function takes as input an `FVector`, an `FRotator`, and an `UnrealEnum`. We can do this very conveniently using the `UnrealObj` class that we're already using in other parts of our code.

```
UnrealObj<FVector> location("Location");
UnrealObj<FRotator> rotation("Rotation");
UnrealObj<FSpEnum> sp_enum("SpEnum");
UnrealObjUtils::setObjectPropertiesFromStrings({&location, &rotation, &sp_enum}, unreal_obj_strings);

// we can wrap this ugly cast operation in a helper function if we find that we are doing it often
UnrealEnum unreal_enum = static_cast<UnrealEnum>(static_cast<std::underlying_type_t<ESpEnum>>(sp_enum.get().SpEnum));

GEngine->UnderlyingUnrealFunction(location.get(), rotation.get(), unreal_enum);
```

In this example, the type of `unreal_obj_strings` is `std::map<std::string, std::string>`, and its key-value pairs should be set as follows.

- `unreal_obj_strings.at("Location")` should be:
 
```
{
   "X": 1.23
   "Y": 4.56
   "Z": 7.89
}
```

- `unreal_obj_strings.at("Rotation")` should be:
 
```
{
   "Roll": 1.23
   "Pitch": 4.56
   "Yaw": 7.89
}
```

- `unreal_obj_strings.at("SpEnum")` should be:
 
```
{
   "SpEnum": "Two"
}
```

I think this approach is preferable to the `MSGPACK` approach for the following reasons.

- Enums can be passed via human-readable strings from Python (e.g., Python code can pass in the string `Two` instead of the integer `2` or a value from a Python enum).
- Enum values and more complicated structs (e.g., `FVector` and `FRotator` in this example) are handled using the exact same code.
- `Unreal::callFunction` and `CppFuncService` already use exactly the same convention for passing Unreal types, so this approach is more consistent with the rest of the code.
- We still need to declare an extra enum in our code, but unlike the `MSGPACK` approach, we can keep this enum synchronized with the underlying Unreal enum without explicitly referring to the integer values (e.g., we don't need to set our `Two = 2` in our code, we can instead set `Two = ugly_cast<ESpEnum>(UnrealEnum::Two)`, which is less brittle).
- We need to declare an extra struct in our code, which is a legitimate downside compared to the `MSGPACK` approach. But it's only a few lines of declarative boilerplate in a header file, so I think it's worth it.

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -36,6 +39,107 @@ struct GameWorldServicePropertyDesc
+    /* Custom function allowing the caller to specific a function to execute post actor construction but before other systems see this actor spawn. */
+    TFunction<void(AActor*)> CustomPreSpawnInitalization;
+};
+
```
I think we can get rid of `set_game_paused` and `open_level` at this point, because a user can accomplish the same result with `get_static_class`, `get_default_object`, and `call_function`.

RachithP replied:
Updated code to do this. It's a bit verbose, but gives a perfect example of how users can compose different function calls to achieve what they want.

mikeroberts3000 replied:
100% agree. You're right that this new style can be bit more verbose on the Python side, and you're right that it'll be a useful reference for users (and ourselves).

For what it's worth, I think some of the added verboseness is unavoidable, given that we're intentionally trying to make our Python interface lower-level, such that it maps more directly to Unreal function calls.

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -6,27 +6,208 @@
-// needed at the SpCore level is sometimes slightly different from the data needed at the SpEngine level, e.g.,
-// PropertyDesc needs to maintain a FProperty pointer but GameWorldServicePropertyDesc doesn't. Third, some
-// data types used in SpCore cannot be sent via RPCLib, e.g., FProperty* in PropertyDesc.
+// We need to include this for UnrealBuildTool to recognize structs with USTRUCT() macro declared in this file.
```
Comment not needed.

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -6,27 +6,208 @@
 
-MSGPACK_ADD_ENUM(EIncludeSuperFlag::Type);
+
+// This corresponds to EIncludeSuperFlag::Type declared in Engine/Source/Runtime/CoreUObject/Public/UObject/Class.h
```
Excellent.

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -6,27 +6,208 @@
+    LOAD_DisableEngineVersionChecks  = static_cast<std::underlying_type_t<ELoadFlags>>(ELoadFlags::LOAD_DisableEngineVersionChecks)
+};
+
+// This struct is intended to be a wrapper for ESpIncludeSuperFlag declared above. Wrapping enums in a struct with USTRUCT() macro
```
Excellent comment. Just write this comment once though.

```
// These USTRUCT types are intended to be wrappers for the UENUM types declared above. Wrapping enums in
// structs like this helps us take advantage of UnrealObj and UnrealObjUtils to pass enums to and from
// Python as a JSON string.

USTRUCT()
struct FSpIncludeSuperFlag
{
    GENERATED_BODY()

    UPROPERTY()
    ESpIncludeSuperFlag Enum;
};

USTRUCT()
struct FSpObjectFlags
{
    GENERATED_BODY()

    UPROPERTY()
    ESpObjectFlags Enum;
};
```

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -6,27 +6,208 @@
+{
+    GENERATED_BODY();
+
+    FSpActorSpawnParameters()
```
Can we do this instead? The following is more more consistent with our existing style. Also I get errors when I have a `;` after `GENERATED_BODY()`. Does that not happen for you?

```
USTRUCT()
struct FSpActorSpawnParameters
{
    GENERATED_BODY()

    UPROPERTY()
    FName Name = Name_None;

    UPROPERTY()
    AActor* Template = nullptr;
}

RachithP replied:
It is weird - I did not get any errors when I used `;` beside `GENERATED_BODY()`.

mikeroberts3000 replied:
Oh, weird, this only seems to happen for `UCLASS` types. I tried to add `;` to `SpCoreActor` and I got the following error:

```
/Users/mroberts/code/github/spear/cpp/unreal_projects/SpearSim/Plugins/SpCore/Source/SpCore/SpCoreActor.h(79): Error: Extra ';' before 'public'
```

But I don't get any errors when I added this `USTRUCT` to the same file, even though the keyword `public` is in the exact same location relative to the `;`. 

```
USTRUCT()
struct FSpEnum
{
    GENERATED_BODY()
public:
    UPROPERTY()
    ESpEnum SpEnum;
};
```

Weird. I propose that we always avoid `GENERATED_BODY();` for consistency, so we don't need to remember this weird inconsistent compile-time behavior.

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -133,7 +320,10 @@ class GameWorldService {
 
         unreal_entry_point_binder->bindFuncUnreal("game_world_service", "find_function_by_name",
-            [this](const uint64_t& uclass, const std::string& name, const EIncludeSuperFlag::Type& include_super_flag) -> uint64_t {
+            [this](const uint64_t& uclass, const std::string& name, const std::map<std::string, std::string>& unreal_obj_strings) -> uint64_t {
```
This is super clean. Considering everything this code needs to do, it doesn't get much cleaner than this.

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -178,8 +368,651 @@ class GameWorldService {
+                SP_ASSERT(location.size() == 3);
+                SP_ASSERT(rotation.size() == 3);
+
+                UnrealObj<FSpActorSpawnParameters> sp_spawn_params("SpawnParams");
```
Let's name this parameter "SpawnParameters" because that is what it's called in the Unreal documentation.

```
UnrealObj<FSpActorSpawnParameters> sp_spawn_params("SpawnParameters");
```

https://docs.unrealengine.com/4.26/en-US/API/Runtime/Engine/Engine/UWorld/SpawnActor/4/

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -178,8 +368,651 @@ class GameWorldService {
+                const std::string& class_name,
+                const uint64_t& outer,
+                const std::string& name,
+                const std::map<std::string, std::string>& unreal_obj_strings,
```
Let's create the convention that `unreal_obj_strings` is always the last parameter (unless there are enum flags, in which case we can have an additional parameter after that of type `const std::map<std::string, std::vector<std::string>>&` called `unreal_flag_strings`.

RachithP replied:
Agreed. In our current functions where we use `unreal_flag_strings`, we do not need an `unreal_obj_strings` argument.

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -178,8 +368,651 @@ class GameWorldService {
+                const uint64_t& in_instance_graph,
+                const uint64_t& external_package) -> uint64_t {
+
+                // TODO: support users to specify multiple object flags
```
Should this say `UnrealObj<FSpObjectFlags>`?

```
// TODO: support multiple object flags
UnrealObj<FSpObjectFlags> sp_object_flags_obj("ObjectFlags");
UnrealObjUtils::setObjectPropertiesFromStrings({&sp_object_flags_obj}, unreal_obj_strings);

ESpObjectFlags sp_object_flags = sp_object_flags.getObj();
EObjectFlags object_flags = static_cast<EObjectFlags>(sp_object_flags.Enum);
```

RachithP replied:
Yes, corrected this typo.

mikeroberts3000 commented on python/spear/game_world_service.py:
```diff
@@ -17,24 +17,40 @@ def open_level(self, level_name):
         return self._rpc_client.call("game_world_service.get_children_components_as_map", actor, include_all_descendants)
 
+    #
+    # Get and Set object properties
```
`Get and set object properties`

mikeroberts3000 commented on python/spear/game_world_service.py:
```diff
@@ -47,39 +63,59 @@ def set_object_properties_from_string_for_uobject(self, uobject, string):
         return self._rpc_client.call("game_world_service.set_property_value_from_string", property_desc, string)
 
+    #
+    # Find and Call functions
```
`Find and call functions`

mikeroberts3000 commented on python/spear/game_world_service.py:
```diff
@@ -47,39 +63,59 @@ def set_object_properties_from_string_for_uobject(self, uobject, string):
+        return self._rpc_client.call("game_world_service.call_function", uobject, ufunction, args)
+
+    #
+    # Find structs by name
```
`Find special structs by name`

mikeroberts3000 commented on python/spear/game_world_service.py:
```diff
@@ -47,39 +63,59 @@ def set_object_properties_from_string_for_uobject(self, uobject, string):
         return self._rpc_client.call("game_world_service.find_special_struct_by_name", name)
 
+    #
+    # Stable Name
```
`Stable name helper functions`

mikeroberts3000 commented on python/spear/game_world_service.py:
```diff
@@ -92,11 +128,282 @@ def get_stable_name_for_actor_component(self, component, include_actor_name):
         return self._rpc_client.call("game_world_service.get_stable_name_for_scene_component", component, include_actor_name)
 
+    #
+    # Get Actor and Component tags
```
`Get actor and component tags`

mikeroberts3000 commented on python/spear/game_world_service.py:
```diff
@@ -92,11 +128,282 @@ def get_stable_name_for_actor_component(self, component, include_actor_name):
         return self._rpc_client.call("game_world_service.get_component_tags", actor)
 
+    #
+    # Get Class
```
`Get class`

mikeroberts3000 commented on python/spear/game_world_service.py:
```diff
@@ -92,11 +128,282 @@ def get_stable_name_for_actor_component(self, component, include_actor_name):
         return self._rpc_client.call("game_world_service.get_class", uobject)
+
+    #
+    # Create Components
```
`Create components`

mikeroberts3000 commented on python/spear/game_world_service.py:
```diff
@@ -92,11 +128,282 @@ def get_stable_name_for_actor_component(self, component, include_actor_name):
+        return self._rpc_client.call("game_world_service.create_scene_component_outside_owner_constructor_from_component", class_name, owner, name)
+
+    #
+    # Create NewObject
```
`Create new object`

mikeroberts3000 commented on python/spear/game_world_service.py:
```diff
@@ -92,11 +128,282 @@ def get_stable_name_for_actor_component(self, component, include_actor_name):
+        return self._rpc_client.call("game_world_service.new_object", class_name, outer, name, {"ObjectFlags": json.dumps({"Enum": flags})}, uobject_template, copy_transients_from_class_defaults, in_instance_graph)
+
+    #
+    # Load Object
```
`Load object`

mikeroberts3000 commented on python/spear/game_world_service.py:
```diff
@@ -92,11 +128,282 @@ def get_stable_name_for_actor_component(self, component, include_actor_name):
+        return self._rpc_client.call("game_world_service.load_object", class_name, outer, name, filename, {"LoadFlags": json.dumps({"Enum": load_flags})}, sandbox, instancing_context)
+
+    #
+    # Get Default Object
```
`Get default object`

mikeroberts3000 commented on python/spear/game_world_service.py:
```diff
@@ -92,11 +128,282 @@ def get_stable_name_for_actor_component(self, component, include_actor_name):
+        return self._rpc_client.call("game_world_service.get_default_object", uclass, create_if_needed)
+
+    #
+    # Get Static Class
```
`Get static class`

mikeroberts3000 commented on python/spear/game_world_service.py:
```diff
@@ -92,11 +128,282 @@ def get_stable_name_for_actor_component(self, component, include_actor_name):
+        return self._rpc_client.call("game_world_service.spawn_actor", class_name, location, rotation, {"SpawnParams": json.dumps(spawn_params)})
+
+    #
+    # Find Actors conditionally and return a list or dict
```
`Find actors conditionally and return a list or dict`

mikeroberts3000 commented on python/spear/game_world_service.py:
```diff
@@ -47,39 +63,59 @@ def set_object_properties_from_string_for_uobject(self, uobject, string):
-        return self._rpc_client.call("game_world_service.find_function_by_name", uclass, name, include_super_flag)
+        return self._rpc_client.call("game_world_service.find_function_by_name", uclass, name, {"IncludeSuperFlag": json.dumps({"Enum": include_super_flag})})
 
     def call_function(self, uobject, ufunction, **kwargs):
```
Does the simpler implementation I proposed in a previous comment not work correctly?

RachithP replied:
It did, updated the code.

mikeroberts3000 commented:
Hey Rachith, you marked this as being ready to review, but there are a still a bunch of unaddressed comments. Was that intended? Did you want me to review again?

RachithP commented:
> Hey Rachith, you marked this as being ready to review, but there are a still a bunch of unaddressed comments. Was that intended? Did you want me to review again?

Hey Mike,
It wasn't intended for another round of review, I changed it as the draft state didn't make sense at this stage of this PR.

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
Hard to skim. Add comments to group the entry points that match the style of those in `Unreal.h` / `UnrealClassRegistrar.h` / `game_world_service.py`.

mikeroberts3000 commented on examples/common/instance_utils.py:
```diff
@@ -19,7 +19,18 @@ def open_level(instance, scene_id, map_id=""):
 
     instance.engine_service.begin_tick()
-    instance.game_world_service.open_level(level_name)
+    gameplay_statics_class = instance.game_world_service.get_static_class(class_name="UGameplayStatics")
```
Very good. This Python interface is starting to get awesome.

mikeroberts3000 commented on examples/common/instance_utils.py:
```diff
@@ -19,7 +19,18 @@ def open_level(instance, scene_id, map_id=""):
+    instance.engine_service.end_tick()
+
+    instance.engine_service.begin_tick()
+    instance.game_world_service.call_function(uobject=gameplay_statics_default_object, ufunction=open_level_func, args={"WorldContextObject": gameplay_statics_default_object, "LevelName": level_name})
```
Why does this function call happen in a separate Unreal frame? Couldn't it happen in the previous frame immediately after `find_function_by_name`? Also, these `WorldContextObject` arguments are handled automatically by Unreal's Blueprint system. We should provide comparable behavior. See this link and search for `WorldContext` if you're curious to find out more,

https://docs.unrealengine.com/4.27/en-US/ProgrammingAndScripting/GameplayArchitecture/Metadata

So, let's change our code as follows,

```
instance.game_world_service.call_function(uobject=gameplay_statics_default_object, ufunction=open_level_func, args={"LevelName": level_name})
```

For this to work, we will need to make the following changes throughout the call stack,

```
def call_function(self, uobject, ufunction, args={}, world_context="WorldContextObject"):
    arg_strings = { arg_name: json.dumps(arg) for arg_name, arg in args.items() }
    return self._rpc_client.call("game_world_service.call_function", uobject, ufunction, arg_strings, world_context)


unreal_entry_point_binder->bindFuncUnreal("game_world_service", "call_function",
    [this](
        const uint64_t& uobject,
        const uint64_t& ufunction,
        const std::map<std::string, std::string>& args,
        const std::string& world_context) -> std::map<std::string, std::string> {

        return Unreal::callFunction(world_, reinterpret_cast<UObject*>(uobject), reinterpret_cast<UFunction*>(ufunction), args, world_context);
    });


// .h
std::map<std::string, std::string> Unreal::callFunction(UWorld* world, UObject* uobject, UFunction* ufunction, const std::map<std::string, std::string>& args, const std::string& world_context = "WorldContextObject")


// .cpp
std::map<std::string, std::string> Unreal::callFunction(UWorld* world, UObject* uobject, UFunction* ufunction, const std::map<std::string, std::string>& args, const std::string& world_context)
{
    SP_ASSERT(world);
    SP_ASSERT(uobject);
    SP_ASSERT(ufunction);

    // ...

    // Set property values.
    for (auto& property_desc : property_descs) {
        std::string property_name = toStdString(property_desc.property_->GetName());

        // If the caller has flagged the property name as the "world context", then set the property using
        // the input world pointer, and make sure the name is not present in args. Strictly speaking, we
        // could assign the world pointer directly to *property_desc.value_ptr, but this could lead to
        // undefined behavior if the property is not actually a pointer. So we assign via setPropertValueFromString(...)
        // because this will avoid undefined behavior.
        if (property_name == world_context) {
            SP_ASSERT(!Std::containsKey(args, property_name));
            setPropertValueFromString(property_desc, Std::toStringFromPtr(world));
        }

        // If the property name is in args, then set it using the string in args, and make make sure the name
        // has not also been flagged as the "world context".
        if (Std::containsKey(args, property_name)) {
            SP_ASSERT(property_name != world_context);
            setPropertyValueFromString(property_desc, args.at(property_name));
        }
    }

    // ...
}
```

RachithP replied:
Wasn't aware that Blueprint function calls automatically handled this `WorldContextParameter` argument for us. Thanks for pointing it out!

mikeroberts3000 commented on python/spear/env.py:
```diff
@@ -27,6 +27,16 @@ def __init__(self, instance, config):
 
         self._ready = False
 
+        self._instance.engine_service.begin_tick()
```
Perfect.

mikeroberts3000 commented on python/spear/env.py:
```diff
@@ -77,13 +87,13 @@ def close(self):
     def begin_tick(self):
         self._instance.engine_service.begin_tick()
-        self._instance.game_world_service.set_game_paused(False)
+        self._instance.game_world_service.call_function(uobject=self.gameplay_statics_default_object, ufunction=self.set_game_paused_func, args={"WorldContextObject": self.gameplay_statics_default_object, "bPaused": False})
```
The Blueprint system handles `WorldContextObject` args automatically and we should too. See comment above.

```
self._instance.game_world_service.call_function(uobject=self.gameplay_statics_default_object, ufunction=self.set_game_paused_func, args={"bPaused": False})
```

mikeroberts3000 commented on python/spear/game_world_service.py:
Great comments, great interface, very easy to skim, very powerful and composable.

mikeroberts3000 commented on examples/mujoco_interop/run.py:
```
name_prefix  = "Meshes/05_chair/"
           ^ too much whitespace
```

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -46,22 +206,16 @@ class GameWorldService {
             });
 
+        //
+        // Find actors unconditionally and return a list or dict
```
`and return an std::vector or std::map`

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -6,27 +6,186 @@
+    ESpawnActorScaleMethod TransformScaleMethod = ESpawnActorScaleMethod::MultiplyWithRoot;
+
+    UPROPERTY()
+    uint8 bNoFail = false;
```
It's weird that we're assigning `false` to a `uint8`. And it's weird that it's a `uint8` but it has a `b` prefix in its name. Shouldn't we make this variable a `bool`?

RachithP replied:
Ah yes, this can be made a bool as it's much easier to read. Unreal sometimes uses `uint8 bVar:1` instead of `bool bVar`.

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -177,9 +375,724 @@ class GameWorldService {
+                    actor_spawn_parameters.NameMode = static_cast<FActorSpawnParameters::ESpawnActorNameMode>(sp_actor_spawn_parameters.NameMode);
+                    actor_spawn_parameters.ObjectFlags = static_cast<EObjectFlags>(sp_actor_spawn_parameters.ObjectFlags);
+
+                    return reinterpret_cast<uint64_t>(
```
```
                    return reinterpret_cast<uint64_t>(UnrealClassRegistrar::spawnActor(class_name, world_, location, rotation, actor_spawn_parameters));
```

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -177,9 +375,724 @@ class GameWorldService {
+        //
+
+        unreal_entry_point_binder->bindFuncUnreal("game_world_service", "spawn_actor",
+            [this](
```
```
            [this](const std::string& class_name, const std::map<std::string, std::string>& unreal_obj_strings) -> uint64_t {
```

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -177,9 +375,724 @@ class GameWorldService {
+                const std::string& class_name,
+                const std::map<std::string, std::string>& unreal_obj_strings) -> uint64_t {
+
+                    UnrealObj<FSpActorSpawnParameters> sp_actor_spawn_parameters_obj("SpawnParameters");
```
Inconsistently indented vs other entry points.

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -177,9 +375,724 @@ class GameWorldService {
+
+        unreal_entry_point_binder->bindFuncUnreal("game_world_service", "find_actors_by_name",
+            [this](const std::string& class_name, const std::vector<std::string>& names,const bool& return_null_if_not_found) -> std::vector<uint64_t> {
+                return Std::reinterpretAsVectorOf<uint64_t>(
```
```
                return Std::reinterpretAsVectorOf<uint64_t>(UnrealClassRegistrar::findActorsByName(class_name, world_, names, return_null_if_not_found));
```

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -177,9 +375,724 @@ class GameWorldService {
+
+        unreal_entry_point_binder->bindFuncUnreal("game_world_service", "find_actors_by_name_as_map",
+            [this](const std::string& class_name, const std::vector<std::string>& names, const bool& return_null_if_not_found) -> std::map<std::string, uint64> {
+                return toUint64(
```
```
                return toUint64(UnrealClassRegistrar::findActorsByNameAsMap(class_name, world_, names, return_null_if_not_found));
```

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -177,9 +375,724 @@ class GameWorldService {
+
+        unreal_entry_point_binder->bindFuncUnreal("game_world_service", "get_component_by_type",
+            [this](const std::string& class_name, const uint64_t& actor, const bool& assert_if_not_found, const bool& assert_if_multiple_found) -> uint64_t {
+                return reinterpret_cast<uint64_t>(
```
```
                return reinterpret_cast<uint64_t>(
                    UnrealClassRegistrar::getComponentByType(class_name, reinterpret_cast<AActor*>(actor), assert_if_not_found, assert_if_multiple_found));
```

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -177,9 +375,724 @@ class GameWorldService {
+
+        unreal_entry_point_binder->bindFuncUnreal("game_world_service", "get_component_by_tag_all",
+            [this](const std::string& class_name, const uint64_t& actor, const std::vector<std::string>& tags, const bool& assert_if_not_found, const bool& assert_if_multiple_found) -> uint64_t {
+                return reinterpret_cast<uint64_t>(
```
```
                return reinterpret_cast<uint64_t>(
                    UnrealClassRegistrar::getComponentByTagAll(class_name, reinterpret_cast<AActor*>(actor), tags, assert_if_not_found, assert_if_multiple_found));
```

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -177,9 +375,724 @@ class GameWorldService {
+
+        unreal_entry_point_binder->bindFuncUnreal("game_world_service", "get_component_by_tag_any",
+            [this](const std::string& class_name, const uint64_t& actor, const std::vector<std::string>& tags, const bool& assert_if_not_found, const bool& assert_if_multiple_found) -> uint64_t {
+                return reinterpret_cast<uint64_t>(
```
```
                return reinterpret_cast<uint64_t>(
                    UnrealClassRegistrar::getComponentByTagAny(class_name, reinterpret_cast<AActor*>(actor), tags, assert_if_not_found, assert_if_multiple_found));
```

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -177,9 +375,724 @@ class GameWorldService {
+
+        unreal_entry_point_binder->bindFuncUnreal("game_world_service", "get_component_by_tag",
+            [this](const std::string& class_name, const uint64_t& actor, const std::string& tag, const bool& assert_if_not_found, const bool& assert_if_multiple_found) -> uint64_t {
+                return reinterpret_cast<uint64_t>(
```
```
                return reinterpret_cast<uint64_t>(
                    UnrealClassRegistrar::getComponentByTag(class_name, reinterpret_cast<AActor*>(actor), tag, assert_if_not_found, assert_if_multiple_found));
```

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -177,9 +375,724 @@ class GameWorldService {
+
+        unreal_entry_point_binder->bindFuncUnreal("game_world_service", "get_component_by_name",
+            [this](const std::string& class_name, const uint64_t& actor, const std::string& name, const bool& assert_if_not_found) -> uint64_t {
+                return reinterpret_cast<uint64_t>(
```
```
                return reinterpret_cast<uint64_t>(
                    UnrealClassRegistrar::getComponentByName(class_name, reinterpret_cast<AActor*>(actor), name, assert_if_not_found));
```

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -177,9 +375,724 @@ class GameWorldService {
+
+        unreal_entry_point_binder->bindFuncUnreal("game_world_service", "get_components_by_name_as_map",
+            [this](const std::string& class_name, const uint64_t& actor, const std::vector<std::string>& names, const bool& return_null_if_not_found) -> std::map<std::string, uint64> {
+                return toUint64(
```
```
                return toUint64(UnrealClassRegistrar::getComponentsByNameAsMap(class_name, reinterpret_cast<AActor*>(actor), names, return_null_if_not_found));
```

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -177,9 +375,724 @@ class GameWorldService {
+
+        unreal_entry_point_binder->bindFuncUnreal("game_world_service", "get_components_by_name",
+            [this](const std::string& class_name, const uint64_t& actor, const std::vector<std::string>& names, const bool& return_null_if_not_found) -> std::vector<uint64_t> {
+                return Std::reinterpretAsVectorOf<uint64_t>(
```
```
                return Std::reinterpretAsVectorOf<uint64_t>(
                    UnrealClassRegistrar::getComponentsByName(class_name, reinterpret_cast<AActor*>(actor), names, return_null_if_not_found));
```

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -177,9 +375,724 @@ class GameWorldService {
+
+        unreal_entry_point_binder->bindFuncUnreal("game_world_service", "find_actor_by_type",
+            [this](const std::string& class_name, const bool& assert_if_not_found, const bool& assert_if_multiple_found) -> uint64_t {
+                return reinterpret_cast<uint64_t>(
```
```
                return reinterpret_cast<uint64_t>(UnrealClassRegistrar::findActorByType(class_name, world_, assert_if_not_found, assert_if_multiple_found));
```

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -177,9 +375,724 @@ class GameWorldService {
+
+        unreal_entry_point_binder->bindFuncUnreal("game_world_service", "find_actor_by_tag_all",
+            [this](const std::string& class_name, const std::vector<std::string>& tags, const bool& assert_if_not_found, const bool& assert_if_multiple_found) -> uint64_t {
+                return reinterpret_cast<uint64_t>(
```
```
                return reinterpret_cast<uint64_t>(UnrealClassRegistrar::findActorByTagAll(class_name, world_, tags, assert_if_not_found, assert_if_multiple_found));
```

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -177,9 +375,724 @@ class GameWorldService {
+
+        unreal_entry_point_binder->bindFuncUnreal("game_world_service", "find_actor_by_tag_any",
+            [this](const std::string& class_name, const std::vector<std::string>& tags, const bool& assert_if_not_found, const bool& assert_if_multiple_found) -> uint64_t {
+                return reinterpret_cast<uint64_t>(
```
```
                return reinterpret_cast<uint64_t>(UnrealClassRegistrar::findActorByTagAny(class_name, world_, tags, assert_if_not_found, assert_if_multiple_found));
```

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -177,9 +375,724 @@ class GameWorldService {
+
+        unreal_entry_point_binder->bindFuncUnreal("game_world_service", "find_actor_by_tag",
+            [this](const std::string& class_name, const std::string& tag, const bool& assert_if_not_found, const bool& assert_if_multiple_found) -> uint64_t {
+                return reinterpret_cast<uint64_t>(
```
```
                return reinterpret_cast<uint64_t>(UnrealClassRegistrar::findActorByTag(class_name, world_, tag, assert_if_not_found, assert_if_multiple_found));
```

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -177,9 +375,724 @@ class GameWorldService {
+
+        unreal_entry_point_binder->bindFuncUnreal("game_world_service", "find_actor_by_name",
+            [this](const std::string& class_name, const std::string& name, const bool& assert_if_not_found) -> uint64_t {
+                return reinterpret_cast<uint64_t>(
```
```
                return reinterpret_cast<uint64_t>(UnrealClassRegistrar::findActorByName(class_name, world_, name, assert_if_not_found));
```

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -142,11 +327,20 @@ class GameWorldService {
+        // Find special structs by name
+        //
+
+
```
Inconsistent whitespace after comment.

mikeroberts3000 commented on python/spear/game_world_service.py:
```diff
@@ -92,11 +111,282 @@ def get_stable_name_for_actor_component(self, component, include_actor_name):
+    # Spawn actor
+    #
+
+    def spawn_actor(self, class_name, location, rotation, spawn_parameters):
```
Need `destroy_actor(actor)` immediately after `spawn_actor(...)` and `destroy_component(component, promote_children)` immediately after `create_scene_component_on_owning_scene_component(...)`.

mikeroberts3000 commented on python/spear/game_world_service.py:
```diff
@@ -92,11 +111,282 @@ def get_stable_name_for_actor_component(self, component, include_actor_name):
+    def create_scene_component_on_component(self, class_name, owner, parent, name):
+        return self._rpc_client.call("game_world_service.create_scene_component_outside_owner_constructor_from_object", class_name, owner, parent, name)
+
+    def create_scene_component_on_owning_scene_component(self, class_name, owner, name):
```
- Need `destroy_actor(self, actor)` immediately after `spawn_actor`.
- Need `destroy_component(self, component, promote_children=False)` immediately after our functions for creating components.
- There is a `K2_DestroyComponent()` `UFUNCTION`, but this function always sets `promote_children` to false, which feels too limiting.
- There is also a `K2_DestroyActor()` `UFUNCTION`, so we could insist that users use this function. But it feels too confusing/asymmetric/inconsistent if we have different approaches for creating and destroying. And it feels too inconsistent if we have a different approaches for actors (i.e., where we could use the `UFUNCTION`) and components (i.e., where we need to use our own function to expose the `promote_children` flag).

mikeroberts3000 commented on python/spear/game_world_service.py:
```diff
@@ -92,11 +111,282 @@ def get_stable_name_for_actor_component(self, component, include_actor_name):
+    # Spawn actor
+    #
+
+    def spawn_actor(self, class_name, location, rotation, spawn_parameters):
```
- Need `destroy_actor(self, actor)` immediately after `spawn_actor`.
- Need `destroy_component(self, component, promote_children=False)` immediately after our functions for creating components.
- There is a `K2_DestroyComponent()` `UFUNCTION`, but this function always sets `promote_children` to false, which feels too limiting.
- There is also a `K2_DestroyActor()` `UFUNCTION`, so we could insist that users use this function. But it feels too confusing/asymmetric/inconsistent if we have different approaches for creating and destroying. And it feels too inconsistent if we have a different approaches for actors (i.e., where we could use the `UFUNCTION`) and components (i.e., where we need to use our own function to expose the `promote_children` flag).

RachithP replied:
I agree, it's only fair to expose symmetrical functions, especially at the user level.

mikeroberts3000 replied:
Agreed. That's why I was fine with taking on a bit of feature creep here (by suggesting we add these `destroy` methods to the PR at the last minute, even though we hadn't discussed them previously). It would be weird if the interfaces were non-symmetric.

mikeroberts3000 commented on cpp/unreal_plugins/SpCore/Source/SpCore/Unreal.cpp:
```diff
@@ -254,6 +254,7 @@ std::string Unreal::getPropertyValueAsString(const Unreal::PropertyDesc& propert
         property_desc.property_->IsA(FFloatProperty::StaticClass())  ||
         property_desc.property_->IsA(FDoubleProperty::StaticClass()) ||
         property_desc.property_->IsA(FStrProperty::StaticClass())    ||
+        property_desc.property_->IsA(FNameProperty::StaticClass())    ||
```
- whitespace not aligned but otherwise correct
- sensibly placed immediately after string, which is a thoughtful choice 👌

mikeroberts3000 commented on cpp/unreal_plugins/SpCore/Source/SpCore/Unreal.cpp:
```diff
@@ -360,6 +361,7 @@ void Unreal::setPropertyValueFromString(const Unreal::PropertyDesc& property_des
         property_desc.property_->IsA(FFloatProperty::StaticClass())  ||
         property_desc.property_->IsA(FDoubleProperty::StaticClass()) ||
         property_desc.property_->IsA(FStrProperty::StaticClass())    ||
+        property_desc.property_->IsA(FNameProperty::StaticClass())   ||
```
- consistently placed in the same order as in the previous usage, again a thoughtful choice 👌

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -177,9 +407,592 @@ class GameWorldService {
+        //
+
+        unreal_entry_point_binder->bindFuncUnreal("game_world_service", "destroy_actor",
+            [this](const uint64_t& actor, const bool& net_force, const bool& should_modify_level) -> bool {
```
I like how the names of the bools match the Unreal documentation.

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -177,9 +407,592 @@ class GameWorldService {
+        // Create component
+        //
+
+        unreal_entry_point_binder->bindFuncUnreal("game_world_service", "create_component_outside_owner_constructor", 
```
I like the decision to keep these entry points named the way they are.

mikeroberts3000 reviewed (approved):
Awesome 🥳

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -132,21 +335,44 @@ class GameWorldService {
+        //
+        // Stable name helper functions
+        //
+
```
A few lines down from this comment:

```
get_stable_name_for_actor_component -> get_stable_name_for_component
```

mikeroberts3000 commented on cpp/unreal_plugins/SpEngine/Source/SpEngine/GameWorldService.h:
```diff
@@ -122,6 +321,10 @@ class GameWorldService {
+        //
+        // Get property values
+        //
+
```
```
        unreal_entry_point_binder->bindFuncUnreal("game_world_service", "get_property_value_as_string",
            [this](const GameWorldServicePropertyDesc& property_desc) -> std::string {
                return Unreal::getPropertyValueAsString(toPropertyDesc(property_desc));
            });

        unreal_entry_point_binder->bindFuncUnreal("game_world_service", "set_property_value_from_string",
            [this](const GameWorldServicePropertyDesc& property_desc, const std::string& string) -> void {
                Unreal::setPropertyValueFromString(toPropertyDesc(property_desc), string);
            });
```

mikeroberts3000 commented on python/spear/game_world_service.py:
```diff
@@ -92,11 +125,282 @@ def get_stable_name_for_actor_component(self, component, include_actor_name):
+    # Destroy actor
+    #
+
+    def destroy_actor(self, actor, net_force=False, should_modify_level=True):
```
Default parameters match the underlying Unreal defaults. Very good.

mikeroberts3000 commented on python/spear/env.py:
```diff
@@ -77,13 +87,13 @@ def close(self):
     def begin_tick(self):
         self._instance.engine_service.begin_tick()
-        self._instance.game_world_service.set_game_paused(False)
+        self._instance.game_world_service.call_function(uobject=self.gameplay_statics_default_object, ufunction=self.set_game_paused_func, args={"bPaused": False})
```
The current name of `self.gameplay_statics_default_object` makes it seem like it's part of this class' public interface, but I don't think it should be. We should change the name as follows.

```
self._gameplay_statics_default_object
```

RachithP replied:
Ah, my bad.  This was an oversight.
